File: demo/src/main/java/com/otaliastudios/cameraview/demo/OptionView.java
package com.otaliastudios.cameraview.demo;


import android.annotation.SuppressLint;
import android.content.Context;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.otaliastudios.cameraview.CameraOptions;
import com.otaliastudios.cameraview.CameraView;

import java.util.ArrayList;

@SuppressLint("ViewConstructor")
public class OptionView<Value> extends LinearLayout implements Spinner.OnItemSelectedListener {

    interface Callback {
        <T> boolean onValueChanged(@NonNull Option<T> option, @NonNull T value, @NonNull String name);
    }

    private Value value;
    private ArrayList<Value> values;
    private ArrayList<String> valuesStrings;
    private Option option;
    private Callback callback;
    private Spinner spinner;

    public OptionView(@NonNull Context context) {
        super(context);
        setOrientation(VERTICAL);
        inflate(context, R.layout.option_view, this);
        ViewGroup content = findViewById(R.id.content);
        spinner = new Spinner(context, Spinner.MODE_DROPDOWN);
        content.addView(spinner);
    }

    public void setHasDivider(boolean hasDivider) {
        View divider = findViewById(R.id.divider);
        divider.setVisibility(hasDivider ? View.VISIBLE : View.GONE);
    }

    public void setOption(@NonNull Option<Value> option, @NonNull Callback callback) {
        this.option = option;
        this.callback = callback;
        TextView title = findViewById(R.id.title);
        title.setText(option.getName());
    }

    @SuppressWarnings("all")
    public void onCameraOpened(CameraView view, CameraOptions options) {
        values = new ArrayList(option.getAll(view, options));
        value = (Value) option.get(view);
        valuesStrings = new ArrayList<>();
        for (Value value : values) {
            valuesStrings.add(option.toString(value));
        }

        if (values.isEmpty()) {
            spinner.setOnItemSelectedListener(null);
            spinner.setEnabled(false);
            spinner.setAlpha(0.8f);
            spinner.setAdapter(new ArrayAdapter(getContext(),
                    R.layout.spinner_text, new String[]{ "Not supported." }));
            spinner.setSelection(0, false);
        } else {
            spinner.setEnabled(true);
            spinner.setAlpha(1f);
            spinner.setAdapter(new ArrayAdapter(getContext(),
                    R.layout.spinner_text, valuesStrings));
            spinner.setSelection(values.indexOf(value), false);
            spinner.setOnItemSelectedListener(this);
        }
    }

    @Override
    public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
        if (!values.get(i).equals(value)) {
            Log.e("ControlView", "curr: " + value + " new: " + values.get(i));
            if (!callback.onValueChanged(option, values.get(i), valuesStrings.get(i))) {
                spinner.setSelection(values.indexOf(value)); // Go back.
            } else {
                value = values.get(i);
            }
        }
    }

    @Override
    public void onNothingSelected(AdapterView<?> adapterView) {}
}

File: demo/src/main/kotlin/com/otaliastudios/cameraview/demo/OptionView.kt
package com.otaliastudios.cameraview.demo

import android.content.Context
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.otaliastudios.cameraview.CameraOptions
import com.otaliastudios.cameraview.CameraView

class OptionView<Value: Any>(context: Context) : LinearLayout(context), AdapterView.OnItemSelectedListener {

    interface Callback {
        fun <T: Any> onValueChanged(option: Option<T>, value: T, name: String): Boolean
    }

    private lateinit var option: Option<Value>
    private lateinit var callback: Callback
    private lateinit var value: Value
    private lateinit var values: List<Value>
    private lateinit var valuesStrings: List<String>

    init {
        orientation = VERTICAL
        View.inflate(context, R.layout.option_view, this)
    }

    val spinner = Spinner(context, Spinner.MODE_DROPDOWN).also {
        val content = findViewById<ViewGroup>(R.id.content)
        content.addView(it)
    }

    fun setHasDivider(hasDivider: Boolean) {
        val divider = findViewById<View>(R.id.divider)
        divider.visibility = if (hasDivider) View.VISIBLE else View.GONE
    }

    fun setOption(option: Option<Value>, callback: Callback) {
        this.option = option
        this.callback = callback
        val title = findViewById<TextView>(R.id.title)
        title.text = option.name
    }

    fun onCameraOpened(view: CameraView, options: CameraOptions) {
        values = option.getAll(view, options).toList()
        value = option.get(view)
        valuesStrings = values.map { option.toString(it) }
        if (values.isEmpty()) {
            spinner.onItemSelectedListener = null
            spinner.isEnabled = false
            spinner.alpha = 0.8f
            spinner.adapter = ArrayAdapter(context, R.layout.spinner_text, arrayOf("Not supported."))
            spinner.setSelection(0, false)
        } else {
            spinner.isEnabled = true
            spinner.alpha = 1f
            spinner.adapter = ArrayAdapter(context, R.layout.spinner_text, valuesStrings)
            spinner.setSelection(values.indexOf(value), false)
            spinner.onItemSelectedListener = this
        }
    }

    override fun onItemSelected(adapterView: AdapterView<*>?, view: View?, i: Int, l: Long) {
        if (values[i] != value) {
            Log.e("ControlView", "curr: " + value + " new: " + values[i])
            if (!callback.onValueChanged(option, values[i], valuesStrings[i])) {
                spinner.setSelection(values.indexOf(value)) // Go back.
            } else {
                value = values[i]
            }
        }
    }

    override fun onNothingSelected(adapterView: AdapterView<*>?) {}
}
